src/vs/workbench/browser/parts/editor/editorActions.ts

```diff
@@ -106,6 +106,71 @@ export class SplitEditorAction extends Action {
 	}
 }
 
+export class JoinEditorsAction extends Action {
+
+	public static ID = 'workbench.action.joinEditors';
+	public static LABEL = nls.localize('joinEditors', "Join Editors");
+
+	constructor(
+		id: string,
+		label: string,
+		@IWorkbenchEditorService private editorService: IWorkbenchEditorService,
+		@IEditorGroupService private editorGroupService: IEditorGroupService
+	) {
+		super(id, label, 'join-editors-action');
+	}
+
+	public run(context?: IEditorContext): TPromise<any> {
+		let editorToJoin: IEditor;
```

Comment:
It looks like you only need `position` and `input` from the `IEditor`. If so, I suggest to completley bypass the `editorService` API and do all of this from the `stacks model` that provides all you need (`inputs`, `groups`, a way to convert from a `group` to a `position`).

The `stacks model` is really the truth of what is going up in the 1-3 `editor groups` and you should be able to do everything with it. The more heavy `IEditor` is the actual instance of control that hosts inputs.